daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -325,22 +307,61 @@ private[lf] object Speedy {
       implicit val loggingContext: LoggingContext,
       /* Compiled packages (Daml-LF ast + compiled speedy expressions). */
       var compiledPackages: CompiledPackages,
-      /* Used when enableLightweightStepTracing is true */
-      var steps: Int,
-      /* Used when enableInstrumentation is true */
-      var track: Instrumentation,
       /* Profile of the run when the packages haven been compiled with profiling enabled. */
-      var profile: Profile,
+      val profile: Profile = new Profile(),
+      val submissionTime: Time.Timestamp,
       /* True if we are running on ledger building transactions, false if we
          are running off-ledger code, e.g., Daml Script or
          Triggers. It is safe to use on ledger for off ledger code but
          not the other way around.
        */
-      val submissionTime: Time.Timestamp,
       val ledgerMode: LedgerMode,
       val disclosureTable: DisclosureTable,
   ) {
 
+    /* The machine control is either an expression or a value. */
+    private[this] var control: Control = Control.Expression(sexpr)
+    /* Frame: to access values for a closure's free-vars. */
+    private[this] var frame: Frame = null
+    /* Actuals: to access values for a function application's arguments. */
+    private[this] var actuals: Actuals = null
+    /* [env] is a stack of temporary values for: let-bindings and pattern-matches. */
+    private[speedy] var env: Env = emptyEnv
+    /* [envBase] is the depth of the temporaries-stack when the current code-context was
+     * begun. We revert to this depth when entering a closure, or returning to the top
+     * continuation on the kontStack.
+     */
+    private[this] var envBase: Int = 0
+    /* Kont, or continuation specifies what should be done next
+     * once the control has been evaluated.
+     */
+    private[speedy] var kontStack: util.ArrayList[Kont] = initialKontStack()
+    /* The last encountered location */
+    private[this] var lastLocation: Option[Location] = None
+    /* Used when enableLightweightStepTracing is true */
+    private[this] var steps: Int = 0
+    /* Used when enableInstrumentation is true */
+    private[this] var track: Instrumentation = Instrumentation()
+
+    private[speedy] def currentControl: Control = control
```

Comment:
What do we achieve with this `currentBlah` = `blah` indirection?

Reply:
Previously, these were all exposed via `var` declarations. This meant that any entity (e.g. some other code running on another thread) with access to the `var` could both read and write the mutable data held here (at any time the entity chose to).

To control this, we need to define an interface layer of code that allows the code owning the `var` to control how and when the mutable data is modified and accessed. The `currentBlah` methods (these would be getter methods in Java land) control how this variable data is read.

In some cases, this alone is sufficient as the result value is immutable (e.g. `currentControl` returns an immutable value), in other cases we need extra engineering work (e.g. `currentFrame`) as we are returning mutable values (which may still be modified by external entities).

Reply:
Thanks for the explanation. Makes sense.